### ISSUE TITLE
Pixelstreaming in Unreal Engine v5.0

### ISSUE THREAD

**Issue by corneel** [Azure/Unreal-Pixel-Streaming#24]
Unreal Engine v5.0 is not generally available yet but the reason why many people would like to try doing pixel streaming with this version of the engine is because of the Nanite and Lumen enhancements that allow projects to scale much better. 

Resources are always a constraint in pixel streaming projects and so I was wondering when you anticipate you will be upgrading your code to accommodate V5.0?

BTW. I was doing some tests with V5.0 and ended up using your 4.27 pixel streaming code with some success but there are still many issues like the matchmaking that don't work correctly.

**Comment by markolofsen** [comment]
@corneel Did you try framework https://github.com/markolofsen/metaeditor for Unreal Engine Pixel Streaming?

**Comment by microsoftman** [comment]
Hi @corneel, thanks for checking out the project. We currently just have been focusing on supporting the 4.27 versions as that is the latest GA one for UE, though I totally understand that with the power of UE5 there are even greater posibilities for incredible graphics with lumen/nanite. This is on the backlog and planned once UE5 becomes GA, as each version seems to break the last, and we want to have the most solid experience. If you happen to have some time to try and export a Pixel Streaming Matchmaker and Signaling Server package from UE5 and do a delta on what major things might need to be tweaked from UE5 versus 4.27.2, I'm happy to review a PR if you can get it working before we do.

**Comment by corneel** [comment]
Thanks, @microsoftman, but I feel this might be more than the average lift for developers that don't have the background in the pixel streaming code. So if I understand it correctly, 5.0 was based on 4.26 but your code was done for 4.27 (did you not know about 5.0?) - so we have a quasi-orphaned code situation. I am using UE5 Early Access because Preview 1 and 2 seem to be way back level.  

That might mean that someone might have to take the pixel streaming code and first backport it to 4.26 and then to 5.0. Either way, I am not sure where Early Access fits it and whether it was based on 4.27. If so, does it perhaps have the same pixel streaming code as 4.27? Who knows? 

In any case, I was able to get an acceptable experience up and running using 5.0 Early Access and 4.27's Javascript code for the Signal and Matchmaker. You are welcome to check it out at http://zfiraverse.com. We even have it running at our partners PureWeb's site - http://ondemand.zfiraverse.com (available 10 am - 10 pm CST). 

I will be honest with you, I am concerned that this great effort Microsoft did with Epic Games for pixel streaming was a flash in the pan - a once-off attempt and that you will not be committed to it in the long term. Not even attempting to investigate what it would take to get the code working with UE 5.0 Preview or Early Access seems to confirm that concern. 

With the rapid increase of popularity in pixel streaming that is happening with the advent of the metaverse one would think that pixel streaming would be a priority for Epic and Microsoft - but perhaps the purchase of Blizzard has switched Microsoft's priorities. Don't believe me - have a look at what @markolofsen shared here above. There are several groups that are muscling into the pixel streaming space - either with rewriting/replacing the primitive Signal and Matchmaking code or by finding some kind of container technology to create an on-demand experience like PureWeb did (with great success btw). 

So to sum up - I wish I had the time to take you up on your challenge but as I am already working more than a hundred hours a week doing a full-time job and building my own business, I don't have the time to help Epic Games fix their own code. Honestly, I would not have minded doing so but I am just slammed at this stage.

**Comment by microsoftman** [comment]
Hi @corneel, I'm happy to announce that in collaboration with Microsoft, Epic has just released an Azure Marketplace solution for Pixel Streaming: [here](https://azuremarketplace.microsoft.com/en-us/marketplace/apps/epicgames.unreal-pixel-streaming?tab=Overview). I want to confim that Microsoft and Epic have a commitment in continuing to make Pixel Streaming in the cloud as seamless as possible for developers. This planned release was mentioned late last year (see video [here](https://www.youtube.com/watch?v=jc8moiBZlM4)), and is now finally live, as mentioned [here](https://forums.unrealengine.com/t/azure-marketplace-unreal-engine-pixel-streaming/507969) in the Unreal forums. It's still yet to be officially announced, but it would be great to start getting your feedback on what should be added next (e.g., UE5 support). This release adds a bunch of new functionality such as lifecycle management and a custom portal for uploading new builds and seeing key metrics, which you can learn more about in the video linked to above. More features are being planned as Epic and MSFT receive initial feedback and gauge adoption.

**Comment by corneel** [comment]
@microsoftman that is great news! I just read through the documentation and I must say, you guys got a few unsolicited, "wow's" from me. Fantastic job! I have some ideas I would like to share with you. Please contact me at corneel.booysen@augmentedenterprise.com.

**Comment by ongk** [comment]
Hi @corneel. Just wanted to add that you should be able to use the packaged bits from UE5 Preview 2 with the Azure Marketplace solution for Pixel Streaming that @microsoftman mentioned. In the step where you need to zip up the package folder, you'll just need to rename the folder from Windows to WindowsNoEditor.

**Comment by corneel** [comment]
@ongk I am using UE5 Early Access 2. There was a notion from some vendors that UE5 was back-level with pixel streaming compared to 4.27. Has this been corrected (and in which version(s))?

**Comment by ongk** [comment]
@corneel Yes, I believe this has been corrected in the UE5 Preview 2 version. I was able to successfully get a basic sample game working from Preview 2.

**Comment by corneel** [comment]
How compatible are Early Access 2 and Preview 2? I tried to open an Early Access project with Preview 1 once and it just reported that all my uassets were unreadable (some message about fortnite). I am not sure if this will be the case in Preview 2 as well.

**Comment by ongk** [comment]
I haven't tried migrating from Early Access to Preview 2 yet, but I have seen some folks in the UE forums talk about issues when upgrading their project to Preview 1. I suspect there will be issues here and there while they work out the kinks in the new versions.

**Comment by corneel** [comment]
Do you know what level the pixel streaming code is in Early Access 2? Is there a way to get this on the latest level just like Preview 2?

**Comment by ongk** [comment]
When I tried out the Early Access 2 version, the matchmaker and signalling server code looked to be an older version. The Preview 2 version looked to me to be 4.27 server code with some minor updates. I'm not sure of the changes in the pixel streaming plugin itself.

**Comment by microsoftman** [comment]
Hi @jeremyxliminal, so Pixel Streaming does work with 5.0, though you need to rename the folder to WindowsNoEditor, as that is what the code in this repo was looking for (UE 4). UE5 creates a Windows folder instead of WindowsNoEditor name.  Or, you can just fork this and tweak the naming in your code too--whatever is easier for you. We will look at adding support to find either or naming in the future, so you won't have to worry about the name change in the future.

**Comment by cristianberbecaru** [comment]
Hi, I'm trying to create an instance but no GPU VMs are available for any region. What should I do ?

**Comment by microsoftman** [comment]
Hello @cristianberbecaru, can you go to the Azure Portal and go to your Subscription->Usage and Quotas section and request the quota that you'd like for the region(s) you will need?  Then please send the ticket # to gamedevvm@microsoft.com and we can help accelerate that request with our capacity team.
